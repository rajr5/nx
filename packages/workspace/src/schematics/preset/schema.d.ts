export interface Schema {
  name: string;
  npmScope?: string;
  style?: string;
  preset: 'empty' | 'angular' | 'react' | 'web-components' | 'full-stack';
}
