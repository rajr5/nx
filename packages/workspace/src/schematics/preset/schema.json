{
  "$schema": "http://json-schema.org/schema",
  "id": "NxWorkspacePreset",
  "title": "Creates applications in a new workspace",
  "type": "object",
  "properties": {
    "preset": {
      "description": "The name of the preset.",
      "type": "string"
    },
    "name": {
      "description": "The name of the application.",
      "type": "string"
    },
    "npmScope": {
      "description": "Npm scope for importing libs.",
      "type": "string"
    },
    "style": {
      "description": "The file extension to be used for style files.",
      "type": "string",
      "default": "css",
      "x-prompt": {
        "message": "Which stylesheet format would you like to use?",
        "type": "list",
        "items": [
          { "value": "css", "label": "CSS" },
          { "value": "scss", "label": "SCSS   [ http://sass-lang.com   ]" },
          { "value": "sass", "label": "SASS   [ http://sass-lang.com   ]" },
          { "value": "less", "label": "LESS   [ http://lesscss.org     ]" },
          { "value": "styl", "label": "Stylus [ http://stylus-lang.com ]" }
        ]
      }
    }
  }
}
