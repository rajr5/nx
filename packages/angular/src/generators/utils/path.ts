import { normalize } from 'path';

export function pathStartsWith(path1: string, path2: string) {
  path1 = normalize(path1).replace(/\\/g, '/');
  path2 = normalize(path2).replace(/\\/g, '/');

  return path1.startsWith(path2);
}
