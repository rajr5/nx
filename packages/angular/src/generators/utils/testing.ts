import type { Tree } from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace();

  const libGenerator = wrapAngularDevkitSchematic('@nrwl/angular', 'library');
  const moduleGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'module'
  );
  const componentGenerator = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'component'
  );

  await libGenerator(tree, { name: libName });

  await componentGenerator(tree, {
    name: 'test-button',
    project: libName,
  });

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, OnInit, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent implements OnInit {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age: number;
  @Input() isOn = false;

  constructor() { }

  ngOnInit() {
  }

}`
  );

  tree.write(
    `libs/${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `libs/${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  await moduleGenerator(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    project: libName,
    path: `libs/${libName}/src/lib/barrel`,
    module: 'barrel',
  });

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `libs/${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  await moduleGenerator(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    project: libName,
    path: `libs/${libName}/src/lib/variable-declare`,
    module: 'variable-declare',
  });

  tree.write(
    `libs/${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create another button in a nested subpath
  await moduleGenerator(tree, {
    name: 'nested',
    project: libName,
    path: `libs/${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    project: libName,
    module: 'nested',
    path: `libs/${libName}/src/lib/nested`,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    project: libName,
  });

  return tree;
}
