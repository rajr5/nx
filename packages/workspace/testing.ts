export {
  createEmptyWorkspace,
  createApp,
  createLib,
  AppConfig,
  getLibConfig,
  getAppConfig
} from './src/utils/testing-utils';
