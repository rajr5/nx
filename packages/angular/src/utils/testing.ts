import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/angular',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/workspace',
  join(__dirname, '../../../workspace/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../storybook/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../migrations.json')
);

/**
 * @deprecated will be removed in v13.
 */
export function runSchematic<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function runMigration<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(schematicName, options, tree)
    .toPromise();
}

/**
 * @deprecated will be removed in v13.
 */
export function runExternalSchematic<SchemaOptions = any>(
  collectionName: string,
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return testRunner
    .runExternalSchematicAsync(collectionName, schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
