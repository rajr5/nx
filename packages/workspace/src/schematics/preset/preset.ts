import {
  chain,
  externalSchematic,
  noop,
  Rule,
  schematic,
  SchematicContext,
  Tree
} from '@angular-devkit/schematics';
import { Schema } from './schema';

import { insert, insertImport, updateJsonInTree } from '../../utils/ast-utils';

import { formatFiles } from '../../utils/rules/format-files';

import * as ts from 'typescript';
import { toFileName } from '@nrwl/workspace/src/utils/name-utils';

export default function(options: Schema): Rule {
  options = normalizeOptions(options);
  return (host: Tree, context: SchematicContext) => {
    return chain([createPreset(options), formatFiles()])(host, context);
  };
}

function createPreset(options: Schema): Rule {
  if (options.preset === 'empty') {
    return noop();
  } else if (options.preset === 'angular') {
    return chain([
      externalSchematic(
        '@nrwl/angular',
        'application',
        {
          name: options.name,
          style: options.style
        },
        { interactive: false }
      ),
      setDefaultCollection('@nrwl/angular')
    ]);
  } else if (options.preset === 'react') {
    return chain([
      externalSchematic(
        '@nrwl/react',
        'application',
        {
          name: options.name,
          style: options.style
        },
        { interactive: false }
      ),
      setDefaultCollection('@nrwl/react')
    ]);
  } else if (options.preset === 'web-components') {
    return chain([
      externalSchematic(
        '@nrwl/web',
        'application',
        {
          name: options.name,
          style: options.style
        },
        { interactive: false }
      ),
      setDefaultCollection('@nrwl/web')
    ]);
  } else {
    return chain([
      externalSchematic(
        '@nrwl/angular',
        'application',
        { name: options.name, style: options.style },
        { interactive: false }
      ),
      externalSchematic(
        '@nrwl/nest',
        'application',
        {
          name: 'api',
          frontendProject: options.name
        },
        { interactive: false }
      ),
      schematic(
        'library',
        { name: 'api-interface', framework: 'none' },
        { interactive: false }
      ),
      setDefaultCollection('@nrwl/angular'),
      connectFrontendAndApi(options)
    ]);
  }
}

function connectFrontendAndApi(options: Schema) {
  const addImportToModule = require('@nrwl/angular/src/utils/ast-utils')
    .addImportToModule;
  return (host: Tree) => {
    host.create(
      'libs/api-interface/src/lib/interfaces.ts',
      `export interface Message { message: string }`
    );
    host.overwrite(
      'libs/api-interface/src/index.ts',
      `export * from './lib/interfaces';`
    );

    const modulePath = `apps/${options.name}/src/app/app.module.ts`;
    const moduleFile = ts.createSourceFile(
      modulePath,
      host.read(modulePath).toString(),
      ts.ScriptTarget.Latest,
      true
    );
    insert(host, modulePath, [
      insertImport(
        moduleFile,
        modulePath,
        'HttpClientModule',
        `@angular/common/http`
      ),
      ...addImportToModule(
        moduleFile,
        `@angular/common/http`,
        `HttpClientModule`
      )
    ]);

    const scope = options.npmScope ? options.npmScope : options.name;
    const style = options.style ? options.style : 'css';
    host.overwrite(
      `apps/${options.name}/src/app/app.component.ts`,
      `import { Component } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Message } from '@${scope}/api-interface';

@Component({
  selector: '${scope}-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.${style}']
})
export class AppComponent {
  hello$ = this.http.get<Message>('/api/hello')
  constructor(private http: HttpClient) {}
}
    `
    );

    host.overwrite(
      `apps/${options.name}/src/app/app.component.spec.ts`,
      `import { Component } from '@angular/core';
import { TestBed, async } from '@angular/core/testing';
import { HttpClientModule } from '@angular/common/http';
import { AppComponent } from './app.component';

describe('AppComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AppComponent],
      imports: [HttpClientModule]
    }).compileComponents();
  }));

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });
});
    `
    );

    host.overwrite(
      `apps/${options.name}/src/app/app.component.html`,
      `<div style="text-align:center">
  <h1>Welcome to ${options.name}!</h1>
  <img
    width="450"
    src="https://raw.githubusercontent.com/nrwl/nx/master/nx-logo.png"
  />
</div>
<div>Message: {{ (hello$|async)|json }}</div>
    `
    );

    host.overwrite(
      `apps/api/src/app/app.controller.ts`,
      `import { Controller, Get } from '@nestjs/common';

import { Message } from '@${scope}/api-interface';

import { AppService } from './app.service';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get('hello')
  getData(): Message {
    return this.appService.getData();
  }
}
    `
    );

    host.overwrite(
      `apps/api/src/app/app.service.ts`,
      `import { Injectable } from '@nestjs/common';
import { Message } from '@${scope}/api-interface';

@Injectable()
export class AppService {
  getData(): Message {
    return { message: 'Welcome to api!' };
  }
}
    `
    );
  };
}

function setDefaultCollection(defaultCollection: string) {
  return updateJsonInTree('angular.json', json => {
    if (!json.cli) {
      json.cli = {};
    }
    json.cli.defaultCollection = defaultCollection;
    return json;
  });
}

function normalizeOptions(options: Schema): Schema {
  options.name = toFileName(options.name);
  return options;
}
