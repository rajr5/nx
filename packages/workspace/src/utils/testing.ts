import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../collection.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationTestRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
