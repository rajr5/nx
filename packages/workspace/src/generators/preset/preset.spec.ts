import { readJson, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { overrideCollectionResolutionForTesting } from '@nrwl/devkit/ngcli-adapter';
import { presetGenerator } from './preset';
import * as path from 'path';

describe('preset', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    overrideCollectionResolutionForTesting({
      '@nrwl/workspace': path.join(
        __dirname,
        '../../../../workspace/collection.json'
      ),
      '@nrwl/angular': path.join(
        __dirname,
        '../../../../angular/collection.json'
      ),
      '@nrwl/linter': path.join(
        __dirname,
        '../../../../linter/collection.json'
      ),
      '@nrwl/nest': path.join(__dirname, '../../../../nest/collection.json'),
      '@nrwl/node': path.join(__dirname, '../../../../node/collection.json'),
      '@nrwl/jest': path.join(__dirname, '../../../../jest/collection.json'),
      '@nrwl/cypress': path.join(
        __dirname,
        '../../../../cypress/collection.json'
      ),
      '@nrwl/express': path.join(
        __dirname,
        '../../../../express/collection.json'
      ),
    });
  });

  afterEach(() => {
    overrideCollectionResolutionForTesting(null);
  });

  it('should create files (preset = angular)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'angular',
      cli: 'nx',
    });
    expect(tree.children('apps/proj')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/app')).toMatchSnapshot();

    expect(
      JSON.parse(tree.read('/workspace.json').toString()).cli.defaultCollection
    ).toBe('@nrwl/angular');
  });

  it('should create files (preset = web-components)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'web-components',
      cli: 'nx',
    });
    expect(tree.exists('/apps/proj/src/main.ts')).toBe(true);
    expect(readJson(tree, '/workspace.json').cli.defaultCollection).toBe(
      '@nrwl/web'
    );
  });

  it('should create files (preset = react)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'react',
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
    });
    expect(tree.exists('/apps/proj/src/main.tsx')).toBe(true);
    expect(readJson(tree, '/workspace.json').cli.defaultCollection).toBe(
      '@nrwl/react'
    );
  });

  it('should create files (preset = next)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'next',
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
    });
    expect(tree.exists('/apps/proj/pages/index.tsx')).toBe(true);
    expect(readJson(tree, '/workspace.json').cli.defaultCollection).toBe(
      '@nrwl/next'
    );
  });

  it('should create files (preset = angular-nest)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'angular-nest',
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
    });

    expect(tree.exists('/apps/proj/src/app/app.component.ts')).toBe(true);
    expect(tree.exists('/apps/api/src/app/app.controller.ts')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/src/lib/api-interfaces.ts')).toBe(
      true
    );
  });

  it('should create files (preset = react-express)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'react-express',
      style: 'css',
      linter: 'eslint',
      cli: 'nx',
    });

    expect(tree.exists('/apps/proj/src/app/app.tsx')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/src/lib/api-interfaces.ts')).toBe(
      true
    );
    expect(tree.exists('/apps/proj/.eslintrc.json')).toBe(true);
    expect(tree.exists('/apps/api/.eslintrc.json')).toBe(true);
    expect(tree.exists('/libs/api-interfaces/.eslintrc.json')).toBe(true);
  });

  it('should create files (preset = express)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: 'express',
      linter: 'eslint',
      cli: 'nx',
    });

    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.exists('apps/proj/.eslintrc.json')).toBe(true);
  });
});
