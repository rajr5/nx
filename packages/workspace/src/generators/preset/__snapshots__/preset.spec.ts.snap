// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset should create files (preset = angular) 1`] = `
Array [
  "tsconfig.editor.json",
  "tsconfig.json",
  "src",
  ".browserslistrc",
  "tsconfig.app.json",
  ".eslintrc.json",
  "jest.config.js",
  "tsconfig.spec.json",
]
`;

exports[`preset should create files (preset = angular) 2`] = `
Array [
  "favicon.ico",
  "index.html",
  "main.ts",
  "polyfills.ts",
  "styles.css",
  "assets",
  "environments",
  "app",
  "test-setup.ts",
]
`;

exports[`preset should create files (preset = angular) 3`] = `
Array [
  "app.module.ts",
  "app.component.css",
  "app.component.html",
  "app.component.spec.ts",
  "app.component.ts",
]
`;
